packages/yarnpkg-pnpify/sources/hoist.ts

```diff
@@ -0,0 +1,283 @@
+  }
+};
+
+/**
+ * Weighs all the packages in the subtree, by adding up own package weight with weights of all
+ * of its direct and transitive dependencies.
+ *
+ * @param tree package tree
+ * @param rootId root package id of the subtree
+ * @param packages package infos
+ * @param nohoist nohoist package ids, that shouldn't be weighed
+ *
+ * @return map of package weights: package id -> total weight
+ */
+const weighPackages = (tree: ReadonlyHoisterPackageTree, rootId: HoisterPackageId, packages: ReadonlyArray<HoisterPackageInfo>, nohoist: ReadonlySet<HoisterPackageId>): WeightMap => {
+  const weights: WeightMap = new Map();
+  const seenIds = new Set<HoisterPackageId>();
+
+  const addUpNodeWeight = (nodeId: HoisterPackageId) => {
+    seenIds.add(nodeId);
+
+    if (!nohoist.has(nodeId)) {
+      weights.set(nodeId, packages[nodeId].weight + (weights.get(nodeId) || 0));
+      for (const depId of tree[nodeId].deps) {
+        if (!seenIds.has(depId)) {
+          addUpNodeWeight(depId);
+        }
+      }
+    }
+  };
+
+  addUpNodeWeight(rootId);
+
+  return weights;
+};
+
+/**
+ * Finds packages that have the max weight among the packages with the same name
+ *
+ * @param weights package weights map: package id -> total weight
+ * @param packages package infos
+ *
+ * @returns package ids with max weights among the packages with the same name
+ */
+const getHeaviestPackages = (weights: ReadonlyWeightMap, packages: ReadonlyArray<HoisterPackageInfo>): Set<HoisterPackageId> => {
+  const heaviestPackages = new Map<HoisterPackageName, {weight: HoisterWeight, pkgId: HoisterPackageId}>();
+  for (const [pkgId, weight] of weights) {
+    const pkgName = packages[pkgId].name;
+    let heaviestPkg = heaviestPackages.get(pkgName);
+    if (!heaviestPkg) {
+      heaviestPkg = {weight, pkgId};
+      heaviestPackages.set(pkgName, heaviestPkg);
+    } else if (weight > heaviestPkg.weight) {
+      heaviestPkg.weight = weight;
+      heaviestPkg.pkgId = pkgId;
+    }
+  }
+
+  const heavyPackageIds = new Set<HoisterPackageId>();
+  for (const {pkgId} of heaviestPackages.values())
+    heavyPackageIds.add(pkgId);
+
+  return heavyPackageIds;
+};
+
+/**
+ * Find the packages that can be hoisted to the subtree root `rootId`.
+ *
+ * @param tree package tree
+ * @param rootId package id that should be regarded as subtree root
+ * @param packages package infos
+ * @param nohoist nohoist package ids
+ */
+const computeHoistCandidates = (tree: HoisterPackageTree, rootId: HoisterPackageId, packages: ReadonlyArray<HoisterPackageInfo>, nohoist: ReadonlySet<HoisterPackageId>): Set<HoisterPackageId> => {
+  // Get current package dependency package names
+  const rootDepNames = new Map<HoisterPackageName, HoisterPackageId>();
+  for (const depId of tree[rootId].deps)
+    rootDepNames.set(packages[depId].name, depId);
+
+  // Weigh all the packages in the subtree
+  const packageWeights = weighPackages(tree, rootId, packages, nohoist);
```

Comment:
Shouldn't we weight only packages that are candidates to hoisting?

Reply:
Done, though the impact is neglible